servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultAsyncContextProvider.java

```diff
@@ -43,7 +52,66 @@ private DefaultAsyncContextProvider() {
     @Nonnull
     @Override
     public ContextMap context() {
-        return CONTEXT_LOCAL.get();
+        final Thread t = Thread.currentThread();
+        if (t instanceof ContextMapHolder) {
+            final ContextMapHolder contextMapHolder = (ContextMapHolder) t;
+            ContextMap map = contextMapHolder.context();
+            if (map == null) {
+                map = newContextMap();
+                contextMapHolder.context(map);
+            }
+            return map;
+        } else {
+            return CONTEXT_THREAD_LOCAL.get();
+        }
+    }
+
+    @Override
+    public ContextMap saveContext() {
+        return context();
+    }
+
+    @Override
+    public Scope attachContext(ContextMap contextMap) {
+        final Thread currentThread = Thread.currentThread();
+        if (currentThread instanceof ContextMapHolder) {
+            final ContextMapHolder asyncContextMapHolder = (ContextMapHolder) currentThread;
+            ContextMap prev = asyncContextMapHolder.context();
+            asyncContextMapHolder.context(contextMap);
+            return prev == null ? Scope.NOOP : () -> detachContext(contextMap, prev);
+        } else {
+            return slowPathSetContext(contextMap);
+        }
+    }
+
+    private static Scope slowPathSetContext(ContextMap contextMap) {
+        ContextMap prev = CONTEXT_THREAD_LOCAL.get();
+        CONTEXT_THREAD_LOCAL.set(contextMap);
+        return () -> detachContext(contextMap, prev);
+    }
+
+    private static void detachContext(ContextMap expectedContext, ContextMap toRestore) {
+        final Thread currentThread = Thread.currentThread();
+        if (currentThread instanceof ContextMapHolder) {
+            final ContextMapHolder asyncContextMapHolder = (ContextMapHolder) currentThread;
+            ContextMap current = asyncContextMapHolder.context();
+            if (current != expectedContext) {
+                LOGGER.warn("Current context didn't match the expected context. current: {}, expected: {}",
```

Comment:
While I really like this logging (is it a practice from OTEL?), I'm worried that if there is a case where we or users messed up we risk overflowing a running system with excess logging and causing an outage (because there could be too many attach/detach operations). Can we either start with `debug` level logging (we can increase level after we test it with a few use-cases) or with a system property that helps to either opt-in or opt-out from this logging?

Reply:
Let's start with debug.